### PR TITLE
Addressing issue #49, offset/length calculation

### PR DESCRIPTION
**Pull request by hugmyndakassi**
- The size of the certificate data WIN_CERTIFICATE::bCertificate is the
  size dwLength minus the offset of the bCertificate in the struct.

### CODE REVIEW

**Comment by hugmyndakassi** [comment]
Actually I didn't get the unit tests to run as of yet. The compiler gave me quite a few errors when trying with CMake and the `-DBUILD_TESTS=1` which was suggested in the README. I need to review how to get this to build successfully so I can run the test suite locally, too.

Either during, but more likely after the weekend.

**Comment by woodruffw** [comment]
ASan doesn't like the changes, JFYI:

```
[  FAILED  ] Auth32Test.read_certs (0 ms)
[ RUN      ] Auth32Test.get_checksums
AddressSanitizer:DEADLYSIGNAL
/home/runner/work/uthenticode/uthenticode/test/uthenticode-test.cpp:44: Failure
=================================================================
==2786==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000020 (pc 0x5630413ec3d8 bp 0x7fffc9dd9050 sp 0x7fffc9dd9010 T0)
==2786==The signal is caused by a READ memory access.
==2786==Hint: address points to the zero page.
```

**Comment by woodruffw** [comment]
Re: https://github.com/trailofbits/uthenticode/pull/53#issuecomment-1024640447: these changes are causing memory corruption, so it's highly likely that your interpretation of the PE's certificate table structure is wrong.

Here's the relevant docs from Microsoft: https://docs.microsoft.com/en-us/windows/win32/debug/pe-format#the-attribute-certificate-table-image-only

Those docs imply that your code is right, but it wouldn't be the first time Microsoft's docs have been wrong.

Here's how WINE does it: https://github.com/wine-mirror/wine/blob/master/dlls/imagehlp/integrity.c#L483-L546

**Comment by hugmyndakassi** [comment]
Looking into this with ASan on my end. For the purpose of testing this it doesn't really matter _what_ comes after the first certificate, as long as the next one is yet again a half-way valid `WIN_CERTIFICATE` struct.

**Comment by hugmyndakassi** [comment]
Okay, so it's definitely wrong to read `length` bytes into the vector at (now) line 403. There are no two ways about it. So on line 404 there should be the `- 8`  after the `length` as you pointed out in your comment. But then it gets strange. Because obviously the offset has at that point already been changed compared to when the loop was entered.

And indeed when single-stepping through the `Debug`-configured `Auth32Test.Certificate_properties` test case into `uthenticode::read_certs()`, without the above I end up with a strange result from `round()`.

Since 1432 (the value of `WIN_CERTIFICATE::dwLength`) is already 8-byte aligned, so by the end of the loop we should either way end up with 1432. But we get 1440.

And now it dawns on me that the main reason for this is that the WINE implementation you referred to does _not_ change `offset` in the course of the loop, but in `uthenticode::read_certs()` offset is bumped up to 8 _prior_ to adding `WIN_CERTIFICATE::dwLength`.

Now what I do not yet fully get is why the adjustment causes an out of bounds error; or rather why that didn't happen before when the full amount of `length` was requested _at `offset`_. Because if I look at the file I can see the whole security directory is 1432 bytes. So correcting for the offset (which was done correctly before), we have 1424 bytes left to read. But since we request `length` which is 1432 at this point it should already fail right there. I don't quite get why this doesn't happen. Still debugging it with the `Debug` config on Linux.

**Comment by hugmyndakassi** [comment]
Ah, my bad. Of course during the initialization of the vector we give the begin and end iterator. So we won't at this point and so correcting the `length` value before this point will wreak havoc.

However, _after_ that point we need to correct for the offset being 8 already. Essentially the offset needs to be reset to the value when entering a given iteration.

**Comment by hugmyndakassi** [comment]
And one more. the end iterator when filling the vector has to be based on the start offset of the _current_ certificate `+ length` added. But currently it takes the beginning of the security directory instead of the current certificate struct.

**Comment by hugmyndakassi** [comment]
I'll amend my change again as to provide a test case with a doctored file.

**Comment by hugmyndakassi** [comment]
Added another bounds check in the latest commit. From my perspective it concludes this pull request.

**Review by woodruffw — APPROVED** [review]

File: src/uthenticode.cpp [hunk 1]
```diff
@@ -389,25 +389,30 @@ std::vector<WinCert> read_certs(peparse::parsed_pe *pe) {
...
+    std::uint16_t revision = *reinterpret_cast<std::uint16_t *>(current_wincert + offset);
     offset += sizeof(revision);
 
-    std::uint16_t type = *reinterpret_cast<std::uint16_t *>(raw_cert_table.data() + offset);
+    std::uint16_t type = *reinterpret_cast<std::uint16_t *>(current_wincert + offset);
     offset += sizeof(type);
 
-    std::vector<std::uint8_t> cert_data(raw_cert_table.data() + offset,
-                                        raw_cert_table.data() + length);
+    // Continue only we can satisfy the length
```
Comment (woodruffw) [inline src/uthenticode.cpp:406]: Thanks for adding this check as well.

**Comment by woodruffw** [comment]
LGTM! Thanks especially for adding a doctored test to exercise the codepath.

**Pull request merged**